default/python/AI/InvasionAI.py

```diff
@@ -76,35 +107,61 @@ def get_invasion_fleets():
+                troops_per_ship = CombatRatingsAI.weight_attack_troops(troops_per_ship, species_troop_grade)
+                if troops_per_ship > best_trooper_count:
+                    best_base_planet = pid2
+                    best_trooper_count = troops_per_ship
+            if best_base_planet != INVALID_ID:
+                foAI.foAIstate.qualifyingTroopBaseTargets.setdefault(pid, [best_trooper_count, INVALID_ID])
+
+        # Pass 2: for each target previously identified for base troopers, check that still qualifies and
+        # check how many base troopers would be needed; if reasonable then queue up the troops and record this in
+        # foAI.foAIstate.qualifyingTroopBaseTargets
+        for pid in foAI.foAIstate.qualifyingTroopBaseTargets.keys():
+            planet = universe.getPlanet(pid)
             if planet and planet.owner == empire_id:
                 del foAI.foAIstate.qualifyingTroopBaseTargets[pid]
-
-        secure_ai_fleet_missions = foAI.foAIstate.get_fleet_missions_with_any_mission_types([MissionType.SECURE])
-        for pid in (set(foAI.foAIstate.qualifyingTroopBaseTargets.keys()) - set(invasion_targeted_planet_ids)):  # TODO: consider overriding standard invasion mission
-            planet = universe.getPlanet(pid)
+                continue
+            if pid in invasion_targeted_planet_ids:  # TODO: consider overriding standard invasion mission
+                continue
             if foAI.foAIstate.qualifyingTroopBaseTargets[pid][1] != -1:
                 reserved_troop_base_targets.append(pid)
                 if planet:
                     all_invasion_targeted_system_ids.add(planet.systemID)
+                # TODO: evaluate changes to situation, any more troops needed, etc.
                 continue  # already building for here
+            _, planet_troops = evaluate_invasion_planet(pid, empire, secure_ai_fleet_missions, False)
             sys_id = planet.systemID
             this_sys_status = foAI.foAIstate.systemStatus.get(sys_id, {})
+            troop_tally = 0
+            for _fid in this_sys_status.get('myfleets', []):
+                troop_tally += FleetUtilsAI.count_troops_in_fleet(_fid)
+            if troop_tally > planet_troops:  # base troopers appear unneeded
+                del foAI.foAIstate.qualifyingTroopBaseTargets[pid]
+                continue
             if (planet.currentMeterValue(fo.meterType.shield) > 0 and
                     this_sys_status.get('myFleetRating', 0) < 0.8 * this_sys_status.get('totalThreat', 0)):
+                # this system not secured, so ruling out invasion base troops for now
+                # don't immediately delete from qualifyingTroopBaseTargets or it will be opened up for regular troops
+                #del foAI.foAIstate.qualifyingTroopBaseTargets[pid]
                 continue
-            loc = foAI.foAIstate.qualifyingTroopBaseTargets[pid][0]
+            loc = foAI.foAIstate.qualifyingTroopBaseTargets[pid][1]
```

Comment:
Here is the problem I see.

At this point, the second entry should always be INVALID_ID as it only changed to a real planet_id when the troopers are enqueued. However, that requires us knowing which planet we should build the bases on. But we always find INVALID_ID and thus will never know the location.

The previous implementation stored the source planet in the first entry and could use that one.
So, considering that you do not use the troop count at the moment, easiest just revert your changes to this structure...